x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java

```diff
@@ -185,7 +221,38 @@ protected void onFinish(ActionListener<Void> listener) {
 
     @Override
     protected SearchRequest buildSearchRequest() {
-        return pivot.buildSearchRequest(getConfig().getSource(), getPosition(), pageSize);
+        SearchRequest searchRequest = new SearchRequest(getConfig().getSource().getIndex());
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        sourceBuilder.aggregation(pivot.buildAggregation(getPosition(), pageSize));
+        sourceBuilder.size(0);
+
+        QueryBuilder pivotQueryBuilder = getConfig().getSource().getQueryConfig().getQuery();
+
+        DataFrameTransformConfig config = getConfig();
+        if (config.getSyncConfig() != null) {
+            if (inProgressOrLastCheckpoint == null) {
+                throw new RuntimeException("in progress checkpoint not found");
+            }
+
+            BoolQueryBuilder filteredQuery = new BoolQueryBuilder().
+                    filter(pivotQueryBuilder).
+                    filter(config.getSyncConfig().getRangeQuery(inProgressOrLastCheckpoint));
+
+            if (changedBuckets != null && changedBuckets.isEmpty() == false) {
```

Comment:
So, if there are *NO* changed buckets, we query the entire timeframe?

Reply:
yes and that will be the case for histogram as histograms are usually small enough anyway. Not sure about date histograms, I haven't looked into those, I think they will add another range query.

What can happen:

```
"terms":{"user_id": []}
```

which is a valid query and as expected returns an empty result set. It can happen if the source index has changed, the indexer got triggered but the _user query_ filters out the changes. This was a bug, but fixed now.

Actually `isEmpty()` should never be false, either `changedBuckets` is `null` or contains keys with empty lists like above.

Good points, needs some more thinking.